### PR TITLE
Fixed errors with `rename()` on Windows when file handle open

### LINKED ISSUE #2: Upgrade 1.1.1 to 1.2.0 break caching feature
### BC Break Report

After running the application on Windows.
```
PHP Warning:  rename(C:\Users\****\AppData\Local\Temp\2\test.php.tmp,test.php): Der Prozess kann nicht auf die Datei zugreifen, da sie von einem anderen Prozess verwendet wird. (code: 32) in D:\Apache_idpm\htdocs\test-config-cache.php on line 18

Warning: rename(C:\Users\****\AppData\Local\Temp\2\test.php.tmp,test.php): Der Prozess kann nicht auf die Datei zugreifen, da sie von einem anderen Prozess verwendet wird. (code: 32) in D:\Apache_idpm\htdocs\test-config-cache.php on line 18
```

|    Q        |   A
|------------ | ------
| laminas-config-aggregator     |1.2.0
| PHP CGI/Fast-CGI     |7.2.23
| Apache     | 2.4
| Windows     | Windows Server 2016

#### Summary

When executing any scripts or running the application the error above appears.
It seems that the function `ConfigAggregator::cacheConfig` generate a file in the `sys_temp_dir` and then try to rename the source file to the desired filename.

I tried to check if a process where using the source file with the service `Process explorer` but no results found.

#### Previous behavior

The file `config-cache.php` is generated in the `data` folder.

#### Current behavior

No file `config-cache.php` is generated and the error above appears.

#### How to reproduce

The `sys_temp_dir` is default to `C:\Users\****\AppData\Local\Temp`
This is a simple php script to reproduce the behaviour on Windows.
```
$cachedConfigFile='test.php';
$tempFile=sys_get_temp_dir() . DIRECTORY_SEPARATOR . basename($cachedConfigFile) . '.tmp';
$fh = fopen($tempFile, 'c');
//chmod($tempFile, '0777');
ftruncate($fh, 0);
fputs($fh,'test');
rename($tempFile, $cachedConfigFile);
fclose($fh);
```

### PR DESCRIPTION
**Pull request by kynx**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

Fixes #2 

@dominicdettabp any chance you could check out my branch and run the unit tests on your Windows setup?

### CODE REVIEW

**Comment by michalbundyra** [comment]
I think it will be better to use https://github.com/webimpress/safe-writer
It was created to be used here and also in the other place in module manager.

Btw. It is already used in https://github.com/Ocramius/ProxyManager

**Comment by dominicdettabp** [comment]
@kynx Well, I can't because the Windows server is used to deploy our application and there is no phpunit installed there. I work on a Mac.
 Last time I created a very simple php script to identity the problem on the Windows server.

**Comment by dominicdettabp** [comment]
@kynx In the end I installed very fast the phpunit version 8 phar to the Windows server and I tried to run your testcases and this is what I see:
```
PHP Fatal error:  Declaration of LaminasTest\ConfigAggregator\ConfigAggregatorTest::setUp() must be compatible with
nit\Framework\TestCase::setUp(): void in C:\Users\dbdidpmadm\Desktop\test-laminas\ConfigAggregatorTest.php on line 2

Fatal error: Declaration of LaminasTest\ConfigAggregator\ConfigAggregatorTest::setUp() must be compatible with PHPUn
ramework\TestCase::setUp(): void in C:\Users\dbdidpmadm\Desktop\test-laminas\ConfigAggregatorTest.php on line 223
```

**Comment by kynx** [comment]
@dominicdettabp Thanks for trying! Yes, looks like this repo uses phpunit 6.3. I'll grab a VM and give it a try when I can.

@michalbundyra Thanks for the pointer to safe-writer. Yeah, it'd be good to have a single way of doing this stuff. But there are a couple of differences:
* the caching code here uses non-blocking locks so multiple processes don't try writing to the same file simultaneously
* safe-writer does the `chmod()` _after_ it's written the cache to the temp file. If the process dies between the two the file will be left with whatever permissions the OS gave it.

I'd want to try to address those before using it.

BTW, it looks like travis is failing on a CS issue - not my fault 😉

File: src/ConfigAggregator.php [hunk 1]
```diff
@@ -298,10 +300,13 @@ private function cacheConfig(array $config, $cachedConfigFile)
             var_export($config, true)
         ));
 
-        rename($tempFile, $cachedConfigFile);
+        // Windows doesn't like renaming an open file, so copy and unlink
+        copy($tempFile, $cachedConfigFile);
```
Comment (Xerkus) [inline src/ConfigAggregator.php]: This is not atomic. Same race condition applies

Comment (kynx) [reply src/ConfigAggregator.php]: Right.

About all I can think of is to either replace the `fopen()`, `fputs()` etc with a `file_put_contents()` or scrap the whole thing, switch to safe-writer and get that setting the permissions before actually writing the file. 

Any suggestions / preferences?

**Comment by diego-sorribas** [comment]
Hi ! Any news on this?

**Pull request closed without merging**